Updating to use c++20

clemahieu opened the pull request:
Fixing various type ambiguities flagged in c++20 mode.

dsiganos commented on nano/lib/numbers.cpp:
```diff
@@ -144,6 +144,11 @@ bool nano::public_key::decode_account (std::string const & source_a)
 
+bool nano::public_key::operator== (nano::public_key const &) const
+{
+	return true;
```
This can't be right.

clemahieu commented:
Looks like boost will need a new version.

Boost asio coroutine support: https://www.boost.org/users/history/version_1_74_0.html
c++20 compiler testing: https://www.boost.org/users/history/version_1_77_0.html

guilhermelawless commented:
Stumbled upon here and remembered this one :)

https://github.com/nanocurrency/nano-node/blob/9bb89ab3232eb603911ae1125a78921ec2d4f591/nano/node/distributed_work_factory.cpp#L57

clemahieu committed:
Consting load_test account_info comparison operator.

clemahieu committed:
Deleting peer_exclusion::item so it can be aggregate initialized.
Using insert rather than emplace since we're inserting an already constructed item.

clemahieu committed:
Changing uint256_union comparison operators to address C++20 overloaded comparison operator ambiguity.

clemahieu committed:
Changing mdb_iterator comparison operators to address C++20 overloaded comparison operator ambiguity.

clemahieu committed:
Using explicit conversion functions to eliminate C++20 comparison operator ambiguities.

clemahieu committed:
Updating compiler version requirement to c++20

thsfs commented:
This change requires updating the documentation website, switching the compiler version from C++17 to C++20.
